test/e2e_node/dra_test.go

```diff
@@ -111,51 +114,387 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
+			pod := createTestObjects(ctx, f.ClientSet, getNodeName(ctx, f), f.Namespace.Name, "draclass", "external-claim", "drapod", 0, true, []string{driverName})
 
 			ginkgo.By("wait for pod to be in Pending state")
 			err := e2epod.WaitForPodCondition(ctx, f.ClientSet, f.Namespace.Name, pod.Name, "Pending", framework.PodStartShortTimeout, func(pod *v1.Pod) (bool, error) {
 				return pod.Status.Phase == v1.PodPending, nil
 			})
 			framework.ExpectNoError(err)
 
-			ginkgo.By("wait for NodePrepareResources call")
-			gomega.Eventually(kubeletPlugin.GetGRPCCalls).WithTimeout(dra.PluginClientTimeout * 2).Should(testdriver.NodePrepareResourcesCalled)
-
 			// TODO: Check condition or event when implemented
 			// see https://github.com/kubernetes/kubernetes/issues/118468 for details
 			ginkgo.By("check that pod is consistently in Pending state")
 			gomega.Consistently(ctx, e2epod.Get(f.ClientSet, pod)).WithTimeout(podInPendingStateTimeout).Should(e2epod.BeInPhase(v1.PodPending),
 				"Pod should be in Pending state as resource preparation time outed")
 		})
+
+		ginkgo.It("must run pod if NodePrepareResources fails and then succeeds", func(ctx context.Context) {
+			ginkgo.By("set failure mode for the plugin")
+			kubeletPlugin.SetFailure(errors.New("Simulated failure"))
+			pod := createTestObjects(ctx, f.ClientSet, getNodeName(ctx, f), f.Namespace.Name, "draclass", "external-claim", "drapod", 0, true, []string{driverName})
+
+			ginkgo.By("wait for pod to be in Pending state")
+			err := e2epod.WaitForPodCondition(ctx, f.ClientSet, f.Namespace.Name, pod.Name, "Pending", framework.PodStartShortTimeout, func(pod *v1.Pod) (bool, error) {
+				return pod.Status.Phase == v1.PodPending, nil
+			})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("wait for NodePrepareResources call to fail")
+			gomega.Eventually(kubeletPlugin.GetGRPCCalls).WithTimeout(dra.PluginClientTimeout * 2).Should(testdriver.NodePrepareResourcesErrored)
+
+			ginkgo.By("set success mode for the plugin")
+			kubeletPlugin.SetFailure(nil)
+
+			ginkgo.By("wait for NodePrepareResources call to succeed")
+			gomega.Eventually(kubeletPlugin.GetGRPCCalls).WithTimeout(dra.PluginClientTimeout * 2).Should(testdriver.NodePrepareResourcesSucceeded)
+
+			ginkgo.By("wait for pod to succeed")
+			err = e2epod.WaitForPodSuccessInNamespace(ctx, f.ClientSet, pod.Name, f.Namespace.Name)
+			framework.ExpectNoError(err)
+		})
+
+		ginkgo.It("must run pod if NodeUnprepareResources fails and then succeeds", func(ctx context.Context) {
+			pod := createTestObjects(ctx, f.ClientSet, getNodeName(ctx, f), f.Namespace.Name, "draclass", "external-claim", "drapod", 10, true, []string{driverName})
+
+			ginkgo.By("wait for pod to run")
+			err := e2epod.WaitForPodRunningInNamespace(ctx, f.ClientSet, pod)
+			framework.ExpectNoError(err)
+
+			ginkgo.By("set failure mode for the plugin")
+			kubeletPlugin.SetFailure(errors.New("Simulated failure"))
```

Comment:
What causes the pod to stop? If it stops by without waiting for anything, then this test is racy (pod might stop and NodeUnprepareResources be called before `SetFailure`).

Reply:
I'm stopping here with the review (no pun intended), but perhaps you can check yourself whether other tests have the same issue (if it is one)?

Reply:
> What causes the pod to stop? If it stops by without waiting for anything, then this test is racy (pod might stop and NodeUnprepareResources be called before SetFailure).

The pod sleeps for 10 secods. I hope this time is enough to set failure mode for UnprepareResource.

Reply:
On a loaded system I have seen failures that could only be explained by code not running for quite some time, longer than 10 seconds.

Instead of depending on timing, can these tests be written so that there is a causal relationship between the different steps?

Reply:
Not sure there is an issue there, but removed waiting for the pod running condition. PTAL

Reply:
If that's not enough, I'll consider introducing separate locks for Prepare* And Unprepare* calls. Just let me know if you want that.

Reply:
This test still has a race.

What I am looking for is a guarantee that this sequence of events occurs regardless of timing:
- pod starts
-  NodePrepareResources succeeded
- pod stops
- NodeUnprepareResources fails at least once
- NodeUnprepareResources succeeds
- pod shutdown completes

Making `kubeletPlugin.SetFailure` specific to certain methods would ensure that. Then the failure mode can be enabled for NodeUnprepareResources before creating the pod.

One nice side effect is that the pod can be allowed to run to completion immediately (no need for a 10 second delay). The test then completes more quickly.

Reply:
done, PTAL